crates/core/tedge_agent/src/entity_manager/server.rs

```diff
@@ -177,27 +180,32 @@ impl EntityStoreServer {
         }
     }
 
-    async fn process_entity_data(&mut self, topic_id: EntityTopicId) {
+    async fn process_entity_data(
+        &mut self,
+        topic_id: EntityTopicId,
+        channel: Channel,
+        message: MqttMessage,
+    ) -> Result<(), entity_store::Error> {
         // if the target entity is unregistered, try to register it first using auto-registration
         if self.entity_store.get(&topic_id).is_none()
             && self.entity_auto_register
             && topic_id.matches_default_topic_scheme()
+            && !message.payload().is_empty()
         {
-            match self.entity_store.auto_register_entity(&topic_id) {
-                Ok(entities) => {
-                    for entity in entities {
-                        let message = entity.to_mqtt_message(&self.mqtt_schema).with_retain();
-                        self.publish_message(message).await;
-                    }
-                }
-                Err(err) => {
-                    error!(
-                        "Failed to auto-register entities for the topic: {} due to: {err}",
-                        &topic_id
-                    );
-                }
+            let entities = self.entity_store.auto_register_entity(&topic_id)?;
+            for entity in entities {
+                let message = entity.to_mqtt_message(&self.mqtt_schema).with_retain();
+                self.publish_message(message).await;
             }
         }
+
+        if let Channel::EntityTwinData { fragment_key } = channel {
+            let fragment_value = serde_json::from_slice(message.payload_bytes())?;
+            let twin_message = EntityTwinMessage::new(topic_id, fragment_key, fragment_value);
+            self.entity_store.update_twin_fragment(twin_message)?;
+        }
```

Comment:
I'm surprised by this bit. This is unrelated to the goal of this PR (clearing retained twin data messages on delete). Ans wasn't twin data already processed by the agent?

Reply:
Yes, I was surprised by that too. This bit got missed during the entity store migration. Writing the test for the bug highlighted this gap. This was never spotted because the entity cache of the mapper was still processing it properly. The `twin_data` maintained by the agent was somewhat unused so far, until now.